dingo/pipe/dag_creator.py

```diff
@@ -78,63 +79,67 @@ def generate_dag(inputs):
+            importance_sampling_generation_node_list = []
+            for idx, trigger_time in enumerate(trigger_times):
+                kwargs = dict(trigger_time=trigger_time, idx=idx, dag=dag)
+                if idx > 0:
+                    # Make all generation nodes depend on the 0th generation node
+                    # Ensures any cached files (e.g. the distance-marginalization
+                    # lookup table) are only built once.
+                    kwargs["parent"] = generation_node_list[0]
+                generation_node = GenerationNode(inputs, importance_sampling=True, **kwargs)
+                importance_sampling_generation_node_list.append(generation_node)
         else:
-            # 4.(b) Recombine jobs into single Result.
-            #       (Automatically calculates evidence.)
-            merge_node = MergeNode(
-                inputs=inputs,
-                parallel_node_list=parallel_node_list,
-                dag=dag,
-            )
-            merged_importance_sampling_node_list.append(merge_node)
+            importance_sampling_generation_node_list = generation_node_list
+
+        #
+        # 4. Importance sample
+        #
+        # If the phase is not present and phase marginalization is not being used, also
+        # sample the phase synthetically. This adds between 1x and 50x to the cost of
+        # importance sampling, depending on the waveform model. Indeed, IMRPhenomXPHM
+        # waveform modes are much more expensive to generate than polarizations.
+
+        merged_importance_sampling_node_list = []
+        parallel_list = get_parallel_list(inputs)
+        all_parallel_node_list = []
+        for sampling_node, generation_node in zip(
+            sampling_node_list, importance_sampling_generation_node_list
+        ):
+            parallel_node_list = []
+            for parallel_idx in parallel_list:
+                importance_sampling_node = ImportanceSamplingNode(
+                    inputs,
+                    sampling_node=sampling_node,
+                    generation_node=generation_node,
+                    parallel_idx=parallel_idx,
+                    dag=dag,
+                )
+                parallel_node_list.append(importance_sampling_node)
+                all_parallel_node_list.append(importance_sampling_node)
+
+            if len(parallel_node_list) == 1:
+                merged_importance_sampling_node_list.append(importance_sampling_node)
+            else:
+                # 4.(b) Recombine jobs into single Result.
+                #       (Automatically calculates evidence.)
+                merge_node = MergeNode(
+                    inputs=inputs,
+                    parallel_node_list=parallel_node_list,
+                    dag=dag,
+                )
+                merged_importance_sampling_node_list.append(merge_node)
+
+    else:
+        merged_importance_sampling_node_list = sampling_node_list
```

Comment:
Consider renaming `merged_importance_sampling_node_list` to `merged_sampling_node_list` or `merged_node_list`.

Reply:
I agree the naming is brutal; I will fix in a later update since I do not want to risk making an error.